src/main.rs

```diff
@@ -0,0 +1,83 @@
+use mdbook::renderer::RenderContext;
+use serde::Deserialize;
+use std::collections::BTreeMap;
+use std::io;
+
+#[derive(Deserialize)]
+struct I18nConfiguration {
+    languages: BTreeMap<String, String>,
+    default_language: Option<String>,
+
+    #[serde(default)]
+    translate_all_languages: bool,
+    #[serde(default)]
+    move_translations_to_html_directory: bool,
+}
+
+fn main() {
+    let mut stdin = io::stdin();
+
+    // Get the configs
+    let ctx = RenderContext::from_json(&mut stdin).unwrap();
+    let i18n_config: I18nConfiguration = ctx
+        .config
+        .get_deserialized_opt("output.i18n-helpers")
+        .unwrap()
+        .unwrap();
+
+    if !i18n_config.translate_all_languages {
+        return;
+    }
+
+    let mut mdbook = mdbook::MDBook::load(&ctx.root).expect("Failed to load book");
+    // Overwrite with current values from stdin.
+    mdbook.book = ctx.book.clone();
+    mdbook.config = ctx.config.clone();
+    mdbook.root = ctx.root.clone();
+
+    let book_config = mdbook
+        .config
+        .get_mut("output.i18n-helpers")
+        .expect("No output.i18n-helpers config in book.toml");
+    // Set translate_all_languages to false for nested builds to prevent infinite recursion.
+    book_config
+        .as_table_mut()
+        .expect("output.i18n-helpers config in book.toml is not a table")
+        .insert(
+            String::from("translate_all_languages"),
+            toml::Value::Boolean(false),
+        );
+
+    let output_directory = ctx.destination.parent().unwrap();
```

Comment:
Here you're crawling out of [`ctx.destination`](https://docs.rs/mdbook/latest/mdbook/renderer/struct.RenderContext.html#structfield.destination), which says

> Where the renderer _must_ put any build artefacts generated.

Instead of relying on `output_directory.join("html")` below, I would suggest the following:

1. Run the `html` renderer somehow and make it output to `ctx.destination`.
2. Run it for each `lang` and make it output to `ctx.destination.join(lang)`.

The advantage: no implicit dependency on people enabling the `html` renderer (I believe it's allowed to disable it).

Reply:
I rewrote that part a bit. That's why we have move_translations_to_html_directory also